File: src/polaplib/scripts/test_fail.py
#!/usr/bin/env python3
# ##############################################################################
# This file is part of polap. GPLv3+ (see top-level license)
# ##############################################################################
# test_fail.py
# Intentional Python failure with full traceback and exit code control.
# Usage: test_fail.py [--message STR] [--exit-code N] [INFILE [OUTFILE]]

import argparse, sys, os, traceback

parser = argparse.ArgumentParser(add_help=True)
parser.add_argument("--message", default="intentional test failure (python)")
parser.add_argument("--exit-code", type=int, default=3, dest="exit_code")
parser.add_argument("infile", nargs="?")
parser.add_argument("outfile", nargs="?")
args = parser.parse_args()

# Optional write
if args.outfile:
    with open(args.outfile, "w") as f:
        f.write("test_fail.py writing before failure\n")


class ExitCodeError(RuntimeError):
    def __init__(self, message, code):
        super().__init__(message)
        self.code = int(code)


# Ensure exit code follows the exception
def _excepthook(exc_type, exc, tb):
    traceback.print_exception(exc_type, exc, tb)
    code = getattr(exc, "code", args.exit_code)
    sys.exit(code)


sys.excepthook = _excepthook


def level3():
    raise ExitCodeError(args.message, args.exit_code)


def level2():
    level3()


def level1():
    level2()


# Emit a little context and then fail
print(
    f"PY test_fail: message='{args.message}' code={args.exit_code} infile='{args.infile or ''}' outfile='{args.outfile or ''}'",
    file=sys.stderr,
)
level1()

File: src/polaplib/scripts/test_fail.R
#!/usr/bin/env Rscript
# ##############################################################################
# This file is part of polap. GPLv3+ (see top-level license)
# ##############################################################################
# test_fail.R
# Intentional R failure with backtrace printing and exit code control.
# Usage: test_fail.R [--message STR] [--exit-code N] [INFILE [OUTFILE]]

args <- commandArgs(trailingOnly = TRUE)

message <- "intentional test failure (R)"
exit_code <- 3L
infile <- ""
outfile <- ""

# Minimal arg parsing
i <- 1L
while (i <= length(args)) {
  a <- args[[i]]
  if (a == "--message" && i + 1L <= length(args)) {
    message <- args[[i + 1L]]
    i <- i + 2L
  } else if (a == "--exit-code" && i + 1L <= length(args)) {
    exit_code <- as.integer(args[[i + 1L]])
    i <- i + 2L
  } else if (a == "-h" || a == "--help") {
    cat("Usage: test_fail.R [--message STR] [--exit-code N] [INFILE [OUTFILE]]\n")
    quit(save = "no", status = 0L)
  } else if (substr(a, 1L, 2L) == "--") {
    stop(sprintf("Unknown option: %s", a))
  } else {
    # positional
    if (identical(infile, "")) infile <- a else if (identical(outfile, "")) outfile <- a
    i <- i + 1L
  }
}

# Optional write
if (nzchar(outfile)) {
  con <- file(outfile, open = "w")
  writeLines("test_fail.R writing before failure", con)
  close(con)
}

# Ensure we get a printed backtrace and controlled exit status in non-interactive runs
mk_handler <- function(code) {
  function() {
    # Base traceback (last error). Depth 2 prints calls; adjust as desired.
    utils::traceback(2)
    # Exit with the requested status
    quit(save = "no", status = code)
  }
}
options(error = mk_handler(exit_code))

third <- function() {
  stop(message, call. = TRUE)
}
second <- function() third()
first <- function() second()

# Emit a little context and then fail
message(sprintf("R test_fail: message='%s' code=%d infile='%s' outfile='%s'", message, exit_code, infile, outfile))
first()
